### PR TITLE
ToricVarieties: Implement structure sheaf

### PR DESCRIPTION
**Pull request by HereAround**
This PR should be rebased once https://github.com/oscar-system/Oscar.jl/pull/1031 is merged. Namely, this other PR changes the `Base.show` for line bundles. Merging both PRs (without the necessary rebase) would therefore lead to failures in the test suite. (-> Changed to draft.)

### CODE REVIEW

**Comment by HechtiDerLachs** [comment]
We're also planning to implement structure sheaves for Schemes; see [this PR](https://github.com/oscar-system/Oscar.jl/pull/936) for a prototype. It is not 100% decided yet which form this takes. But I would be happy to hear what exactly you intend to do here, so that we can entangle and fuse from an early stage on (and also avoid name conflicts).

So far, our idea is the following: 

 * on a scheme X there is a struct `SheafOfRings`.
 * given an open set U in X and an instance R of `SheafOfRings`, calling R(U) should return a ring `<:AbstractAlgebra.Ring`.
 * restrictions of ring elements must be implemented.

Depending on the type of U, it is possible to return rings of different type via dispatch. For instance, whenever U is an affine open subset, we can chose to return the associated quotient of a polynomial ring. In the case of toric varieties, this would be cooked up from your `toric_ideal` method. 

But your concepts for structure sheaves are probably far from being so concrete in terms of algebra, correct?

**Comment by fingolfin** [comment]
PR #1031 was merged, so feel free to rebase.

**Comment by HereAround** [comment]
Yes, I did not intend to go much into these algebraic directions. Here is a loose collection of what I had in mind. Most of the following will not be covered in this PR, some of it is not my priority and other parts are very involved and complicated and will therefore take very, very long (e.g. the categorical Serre quotient). So here is what I see towards coherent sheaves and their algebraic counterparts on toric varieties:
- Line bundles as "lists of integers". This is more or less what we currently have. For most of what I currently want to do with line bundles, this is sufficient.
- Line bundles as sheafification of (shifts of) the cox ring S. This allows to compute a basis of global sections from all polynomials of vanishing degree. Alternatively (and more toric), we can compute certain Laurent monomials.
- Vector bundles as sheafifications of free and graded S-modules.
- Coherent sheaves as sheafifications of f.p. graded S-modules. In particular, maps among vector and line bundles via maps of f.p. graded S-modules.
- Normal and tangent bundles, resolutions (in particular Koszul resolutions) for coherent sheaves on complete intersection subvarieties.
- ...

Restrictions to affine open sets only come to my mind at a very late stage. Namely, this allows to tell via a categorical Serre quotient (which we of course do not have in `Oscar` yet) if two different f.p. graded S-modules encode the same coherent sheaf.

Long story short: We could at this point attach the Cox ring as attribute to the structure sheaf. From this you could in principle compute restrictions to affine open subsets. At the current stage however, I see only very few benefits.

File: src/ToricVarieties/LineBundles/constructors.jl [hunk 1]
```diff
@@ -48,7 +48,37 @@ end
 
 
 ########################
-# 3: Display
+# 3: Special constructors
+########################
+
+# The following is both a special constructor of a toric line bundle and an attribute of a toric variety.
+# I consider as the latter, hence the function name is not capitalized. Also, I cache the result.
```
Comment (fingolfin) [inline src/ToricVarieties/LineBundles/constructors.jl]: General remark: our users won't be Julia experts in general (hopefully -- if only Julia experts end up using Oscar, we did something wrong). Therefore, we can't expect them to understand the subtle differences between a constructor and a "regular" function that just happens to return a "new" object.

Thus we decided long ago that ideally all constructors should also be accessible under a suitable `snake_case` name. Indeed, even for Julia experts it can be difficult to guess whether to use, say, `StructureSheaf` or `structure_sheaf` in case both exist.

**Review by fingolfin — APPROVED** [review]

**Comment by fingolfin** [comment]
This needs a rebase

**Comment by HereAround** [comment]
Thx @fingolfin . This method can now be called via `StructureSheaf` and `structure_sheaf`. I have also just rebased this.

**Review by lkastner — APPROVED** [review]

**Comment by HereAround** [comment]
Can this be merged?

**Comment by fingolfin** [comment]
I am happy this was merged, but I still wonder about the name `StructureSheaf`. As far as I can this isn't a type, so why do we use it at all?

**Comment by HereAround** [comment]
I am not entirely certain if I follow. Still, here is an attempt towards answering your question.

To my knowledge, `structure_sheaf` is (as of now) indeed not an `Oscar` data type. Rather, it refers to a special line bundle on a toric variety, specifically the trivial one. As such, I was torn between implementing
1. an attribute of a toric variety (-> `structure_sheaf`),
2. a special constructor for a line bundle (-> `StructureSheaf`).

In my first draft, I used `structure_sheaf`. Later, I followed your suggestion above and supported `StructureSheaf` also.

Does that answer your question? (Sorry if I missed your point. In this case, please elaborate more.)
